core/domain/html_cleaner.py

```diff
@@ -364,9 +376,24 @@ def convert_to_text_angular(html_data):
     # Ensure that p tag is not wrapped in p tag.
     for p in soup.findAll('p'):
         if p.parent.name == 'p':
-            p.unwrap()
+            contents = p.parent.contents
+            length_of_contents = len(contents)
+            index = 0
+            while index < length_of_contents:
+                content = contents[index]
+                new_p = soup.new_tag('p')
+                if content.name != 'p':
+                    while content.name != 'p':
```

Comment:
Can we simplify this? For example I'm not sure why the if statement in the line above is needed, and I'm also a bit uneasy about all the mutating of length_of_contents and index.

Instead, would prefer an approach that's something like this:

```
    if p.parent.name == 'p':
        child_tags = p.parent.contents
        new_child_tags = []
        for child_tag in child_tags:
            # change child_tag as needed (wrap it in p, or unwrap it, etc.) and append it to new_child_tags
        # Update p.parent to contain the new_child_tags instead
        # Unwrap p.parent
```

For loops are better than while loops because it's clear they will terminate, and also the logic in this approach seems similar. The only issue is that I don't know if the above actually does what you're intending to do, because I'm having a bit of difficulty following the original code. What are your thoughts?

Reply:
What I want to do is as follows:
If any p tag is a child of p tag then all components which are not independent parents within the parent tag should be wrapped in new p tags. When a child tag is wrapped, the content list is automatically updated. 

The if statement is needed because if a child is not in a p tag, then in that case only we need to check next child i.e. the continuous children who are not wrapped in a p will be wrapped together.

If the child is already in a p tag, no further check is required.

while loop is needed here since the index has to be updated according to the change in content and the next content will be accessed according to that index.

In the approach mentioned by you, I would need to check consecutive child tags as well and then the for loop will not work there. So, I need to use while loop there and the child_tags list will be updated when two children as clubbed as a single child after wrapping.

Reply:
Thanks for the explanation. I'm still trying to parse this, so a few questions:

> all components which are not independent parents

What is an "independent" parent? (I don't think I've come across this term.)

> ... because if a child is not in a p tag

Do you mean "if a child is not a p tag"? I thought all the children were, by definition, children of a p tag.

-----------------

If I understand you correctly, what you want to do is:
- Create a list of children of the parent p tag
- For children which are already p tags, leave them as they are
- Find maximally-contiguous "substrings" that are composed entirely of not-p tags (so, they could be text nodes, or other HTML tags), and wrap each of these "substrings" into one big p tag.

Is that correct?

Reply:
Yes that is right.

Reply:
Also, if my understanding is correct, you can still use a for loop, I think. Something like this (not tested, and I'm guessing at the BeautifulSoup API):

```
    if p.parent.name == 'p':
        child_tags = p.parent.contents
        new_child_tags = []
        buffer = []

        # Leave existing p-tags as they are; wrap maximally-contiguous substrings of
        # non-p-tags into a single p-tag.
        for child_tag in child_tags:
            if child_tag.name == 'p':
                if buffer:
                    new_child_tags.append(buffer.wrap('p'))
                    buffer = []
                new_child_tags.append(child_tag)
           else:
                buffer.append(child_tag)
 
            if buffer:
                new_child_tags.append(buffer.wrap('p'))
                buffer = []
 ```

Reply:
I don't get what you mean to do by buffer.wrap. `wrap` can be applied to a tag type and buffer is a list, so does that imply wrapping all tags in buffer?

Reply:
Yes, sorry. By wrap() i meant whatever the command was to concat all tags in buffer and wrap them in a single p-tag.

Reply:
Thanks @bansalnitish for the explanation. Will think about this and follow up, but merging this in the meantime.